### PR TITLE
Add PKCS7_sign, PKCS7_final, SMIME_write_PKCS7 to compatibility layer

### PR DESCRIPTION
**Pull request by cconlon**
# Description

This PR adds the following APIs to the OpenSSL compatibility layer.  PKCS7_sign and SMIME_write_PKCS7 are used in OpenSplice.

- `PKCS7_sign`
- `PKCS7_final`
- `SMIME_write_PKCS7`

Also adds support for the following flags used by these functions as well as PKCS7_verify(): `PKCS7_TEXT`, `PKCS7_DETACHED`, `PKCS7_STREAM`.

# Testing

- API tests added to api.c for the above functions
- Interop tested against OpenSSL using simple test apps based on [OpenSSL SMIME test apps](https://github.com/openssl/openssl/tree/master/demos/smime)

# Checklist

 - [x] added tests
 - [ ] updated/added doxygen
 - [ ] updated appropriate READMEs
 - [ ] Updated manual and documentation

### CODE REVIEW

File: src/ssl.c [hunk 1]
```diff
@@ -20598,7 +20598,7 @@ size_t wolfSSL_get_client_random(const WOLFSSL* ssl, unsigned char* out,
 
 
 #if (defined(KEEP_PEER_CERT) && defined(SESSION_CERTS)) || \
-                                   (defined(OPENSSL_ALL) && defined(HAVE_PKCS7))
+    (defined(OPENSSL_ALL) && defined(SESSION_CERTS))
```
Comment (haydenroche5) [inline src/ssl.c:20601]: Can you explain the change from `HAVE_PKCS7` to `SESSION_CERTS`? Also, the `#endif` comment doesn't match now:

```
#endif /* (KEEP_PEER_CERT && SESSION_CERTS) || (OPENSSL_ALL && HAVE_PKCS7) */
```

File: src/ssl.c [hunk 2]
```diff
@@ -62668,12 +62678,7 @@ PKCS7* wolfSSL_d2i_PKCS7_ex(PKCS7** p7, const unsigned char** in, int len,
     if ((pkcs7 = (WOLFSSL_PKCS7*)wolfSSL_PKCS7_new()) == NULL)
         return NULL;
 
-    if (GetSequence(*in, &idx, &pkcs7->len, len) < 0) {
-        wolfSSL_PKCS7_free((PKCS7*)pkcs7);
-        return NULL;
-    }
-    pkcs7->len += idx;
-
+    pkcs7->len = len;
```
Comment (haydenroche5) [inline src/ssl.c:62681]: Not sure if necessary, but I wonder if we should check that `len` is > 0.

File: src/ssl.c [hunk 3]
```diff
@@ -62753,34 +62758,33 @@ WOLFSSL_STACK* wolfSSL_PKCS7_to_stack(PKCS7* pkcs7)
...
+    signers = wolfSSL_sk_X509_new();
     if (signers == NULL)
         return NULL;
 
-    signers->num = 1;
-    signers->data.x509 = (WOLFSSL_X509*)XMALLOC(sizeof(WOLFSSL_X509), NULL,
-        DYNAMIC_TYPE_X509);
-    if (signers->data.x509 == NULL) {
-        XFREE(signers, NULL, DYNAMIC_TYPE_X509);
+    if (wolfSSL_d2i_X509(&x509, (const byte**)&p7->pkcs7.singleCert,
```
Comment (haydenroche5) [inline src/ssl.c:62780]: Nice refactor of this function.

File: src/ssl.c [hunk 4]
```diff
@@ -62862,6 +62866,7 @@ int wolfSSL_i2d_PKCS7(PKCS7 *p7, unsigned char **out)
             WOLFSSL_MSG("malloc error");
             goto cleanup;
         }
+        XMEMSET(output, 0, len);
```
Comment (haydenroche5) [inline src/ssl.c:62869]: Why is this needed?

File: src/ssl.c [hunk 5]
```diff
@@ -62915,13 +62920,337 @@ int wolfSSL_i2d_PKCS7_bio(WOLFSSL_BIO *bio, PKCS7 *p7)
     return ret;
 }
 
+/**
+ * Creates and returns a PKCS7 signedData structure.
+ *
+ * Inner content type is set to DATA to match OpenSSL behavior.
+ *
+ * signer - certificate to sign bundle with
+ * pkey     - private key matching signcert
```
Comment (haydenroche5) [inline src/ssl.c:62929]: What's signcert? Should it be signer?

File: src/ssl.c [hunk 6]
```diff
@@ -62915,13 +62920,337 @@ int wolfSSL_i2d_PKCS7_bio(WOLFSSL_BIO *bio, PKCS7 *p7)
     return ret;
 }
 
+/**
+ * Creates and returns a PKCS7 signedData structure.
+ *
+ * Inner content type is set to DATA to match OpenSSL behavior.
+ *
+ * signer - certificate to sign bundle with
```
Comment (haydenroche5) [inline src/ssl.c:62928]: Align the "-".

File: src/ssl.c [hunk 7]
```diff
@@ -62915,13 +62920,337 @@ int wolfSSL_i2d_PKCS7_bio(WOLFSSL_BIO *bio, PKCS7 *p7)
...
+ * pkey     - private key matching signcert
+ * certs    - optional additional set of certificates to include
+ * in       - input data to be signed
+ * flags    - optional set of flags to control sign behavior
+ *
+ *    PKCS7_BINARY - Do not translate input data to MIME canonical
+ *                   format (\r\n line endings), thus preventing corruption of
+ *                   binary content.
+ *    PKCS7_TEXT - Prepend MIME headers for text/plain to content.
```
Comment (haydenroche5) [inline src/ssl.c:62937]: Align these "-", too.

File: src/ssl.c [hunk 8]
```diff
@@ -62915,13 +62920,337 @@ int wolfSSL_i2d_PKCS7_bio(WOLFSSL_BIO *bio, PKCS7 *p7)
...
+        WOLFSSL_MSG("Bad function arg, pkey is NULL or incomplete");
+        err = 1;
+    }
+
+    if ((err == 0) && (in == NULL) && !(flags & PKCS7_STREAM)) {
+        WOLFSSL_MSG("input data required unless PKCS7_STREAM used");
+        err = 1;
+    }
+
+    if ((err == 0) && ((p7 = (WOLFSSL_PKCS7*)wolfSSL_PKCS7_new()) == NULL)) {
```
Comment (haydenroche5) [inline src/ssl.c:62985]: Interesting -- how does this cast work? It seems like `WOLFSSL_PKCS7` and `PKCS7` are different types. I see this has been done elsewhere, too, but I'm just curious.

File: src/ssl.c [hunk 9]
```diff
@@ -62915,13 +62920,337 @@ int wolfSSL_i2d_PKCS7_bio(WOLFSSL_BIO *bio, PKCS7 *p7)
...
+    if (err == 0) {
+        p7->pkcs7.privateKey = (byte*)pkey->pkey.ptr;
+        p7->pkcs7.privateKeySz = pkey->pkey_sz;
+        p7->pkcs7.contentOID = DATA;  /* inner content default is DATA */
+        p7->pkcs7.hashOID = SHA256h;  /* default to SHA-256 hash type */
+        p7->type = SIGNED_DATA;       /* PKCS7_final switches on type */
+    }
+
+    /* add additional chain certs if provided */
+    if (err == 0) {
```
Comment (haydenroche5) [inline src/ssl.c:63009]: I think you can remove this outer if, since the while loop checks `err`.

File: src/ssl.c [hunk 10]
```diff
@@ -62915,13 +62920,337 @@ int wolfSSL_i2d_PKCS7_bio(WOLFSSL_BIO *bio, PKCS7 *p7)
...
+#endif
+    XMEMSET(line, 0, MAX_MIME_LINE_LEN);
+
+    while ((lineLen = wolfSSL_BIO_gets(in, line, (int)sizeof(line))) > 0) {
+
+        if (line[lineLen - 1] == '\r' || line[lineLen - 1] == '\n') {
+            canonLineLen = (word32)lineLen;
+            if ((canonLine = wc_MIME_canonicalize(
+                                line, &canonLineLen)) == NULL) {
+                ret = 1;
```
Comment (haydenroche5) [inline src/ssl.c:63094]: Should this and the other `ret = 1`s below be `ret = -1`?

File: src/ssl.c [hunk 11]
```diff
@@ -62915,13 +62920,337 @@ int wolfSSL_i2d_PKCS7_bio(WOLFSSL_BIO *bio, PKCS7 *p7)
...
+
+    return ret;
+}
+
+#endif /* HAVE_SMIME */
+
+/**
+ * Finalize PKCS7 structure, currently supports signedData only.
+ *
+ * Does not do generation of final bundle (ie: signedData), but finalizes
```
Comment (haydenroche5) [inline src/ssl.c:63134]: Does not do generation of -> Does not generate.

File: src/ssl.c [hunk 12]
```diff
@@ -62915,13 +62920,337 @@ int wolfSSL_i2d_PKCS7_bio(WOLFSSL_BIO *bio, PKCS7 *p7)
...
+ *
+ * pkcs7 - initialized PKCS7 structure, populated with signer, etc
+ * in    - input data
+ * flags - flags to control PKCS7 behavior. Other flags except those noted
+ *         below are ignored:
+ *
+ *    PKCS7_BINARY - Do not translate input data to MIME canonical
+ *                   format (\r\n line endings), thus preventing corruption of
+ *                   binary content.
+ *    PKCS7_TEXT - Prepend MIME headers for text/plain to content.
```
Comment (haydenroche5) [inline src/ssl.c:63145]: Align "-".

File: src/ssl.c [hunk 13]
```diff
@@ -62915,13 +62920,337 @@ int wolfSSL_i2d_PKCS7_bio(WOLFSSL_BIO *bio, PKCS7 *p7)
...
+ *
+ * Returns 1 on success, 0 on error
```
Comment (haydenroche5) [inline src/ssl.c:63147]: Could use `WOLFSSL_SUCCESS` and `WOLFSSL_FAILURE`.

File: src/ssl.c [hunk 14]
```diff
@@ -62915,13 +62920,337 @@ int wolfSSL_i2d_PKCS7_bio(WOLFSSL_BIO *bio, PKCS7 *p7)
...
+    }
+
+    if (ret == 1) {
+        if ((data = wolfSSL_BIO_new(wolfSSL_BIO_s_mem())) == NULL) {
+            WOLFSSL_MSG("Error in wolfSSL_BIO_new");
+            ret = 0;
+        }
+    }
+
+    /* append Content-Type header if PKCS7_TEXT */
```
Comment (haydenroche5) [inline src/ssl.c:63172]: append -> prepend (I think)

File: src/ssl.c [hunk 15]
```diff
@@ -62915,13 +62920,337 @@ int wolfSSL_i2d_PKCS7_bio(WOLFSSL_BIO *bio, PKCS7 *p7)
...
+    if ((ret == 1) && (flags & PKCS7_TEXT)) {
+        if (wolfSSL_BIO_write(data, contTypeText,
+                              (int)XSTRLEN(contTypeText)) < 0) {
+            WOLFSSL_MSG("Error appending Content-Type header");
```
Comment (haydenroche5) [inline src/ssl.c:63176]: appending -> prepending

File: src/ssl.c [hunk 16]
```diff
@@ -62915,13 +62920,337 @@ int wolfSSL_i2d_PKCS7_bio(WOLFSSL_BIO *bio, PKCS7 *p7)
...
+    if (ret == 1) {
+        if (flags & PKCS7_BINARY) {
+            /* no CRLF conversion, direct copy content */
+            if ((memSz = wolfSSL_BIO_get_mem_data(in, &mem)) < 0) {
+                ret = 0;
+            }
+            else if (wolfSSL_BIO_write(data, mem, memSz) < 0) {
+                ret = 0;
+            }
+            mem = NULL;
```
Comment (haydenroche5) [inline src/ssl.c:63191]: Are these resets needed?

**Review by haydenroche5 — CHANGES_REQUESTED** [review]
May be a good idea for someone with PKCS7 and/or S/MIME experience to review this, too.

File: src/ssl.c (hunk 1 above)
Comment (cconlon) [reply src/ssl.c:20601]: DecodeToX509() is only called in two places in ssl.c.  Once in the case of `KEEP_PEER_CERT & SESSION_CERTS` and once in the case of `OPENSSL_EXTRA & SESSION_CERTS`.  I'll push a change that corrects `OPENSSL_ALL` to `OPENSSL_EXTRA` and adjusts the endif comment.

File: src/ssl.c (hunk 2 above)
Comment (cconlon) [reply src/ssl.c:62681]: Sure, I'll add this check.

File: src/ssl.c (hunk 4 above)
Comment (cconlon) [reply src/ssl.c:62869]: Likely not necessary, was just being cautious before passing into wc_PKCS7_EncodeSignedData().  I think we can remove since we're allocating the expected size.

File: src/ssl.c (hunk 5 above)
Comment (cconlon) [reply src/ssl.c:62929]: Good catch, will correct.

File: src/ssl.c (hunk 6 above)
Comment (cconlon) [reply src/ssl.c:62928]: Yep, will fix, thanks

File: src/ssl.c (hunk 7 above)
Comment (cconlon) [reply src/ssl.c:62937]: Sure, will correct

File: src/ssl.c (hunk 8 above)
Comment (cconlon) [reply src/ssl.c:62985]: I was following our existing pattern here.  I didn't check with Juliusz, but my best guess was that OpenSSL-consuming apps expect the returned struct to be `PKCS7`.  Since we already have a `PKCS7` struct of our own, he introduced a `WOLFSSL_PKCS7` which holds an inner `PKCS7`.  I think this should work as long as we **always** cast it back to the right `WOLFSSL_PKCS7` type before accessing/referencing members inside, but it does seem fragile.

File: src/ssl.c (hunk 9 above)
Comment (cconlon) [reply src/ssl.c:63009]: Good catch, will remove.

File: src/ssl.c (hunk 10 above)
Comment (cconlon) [reply src/ssl.c:63094]: Thanks, will fix!

File: src/ssl.c (hunk 11 above)
Comment (cconlon) [reply src/ssl.c:63134]: More concise, will adjust.

File: src/ssl.c (hunk 12 above)
Comment (cconlon) [reply src/ssl.c:63145]: Thanks, will align.

File: src/ssl.c (hunk 14 above)
Comment (cconlon) [reply src/ssl.c:63172]: Technically we are prepending, sounds correct.

File: src/ssl.c (hunk 13 above)
Comment (cconlon) [reply src/ssl.c:63147]: I left this as 1/0 in the description to match OpenSSL's man page expected returns.

File: src/ssl.c (hunk 15 above)
Comment (cconlon) [reply src/ssl.c:63176]: Will adjust, thanks.

File: src/ssl.c (hunk 16 above)
Comment (cconlon) [reply src/ssl.c:63191]: Looking this over again, I don't believe so.  We reuse below, but they should get reset at that point.

**Comment by cconlon** [comment]
@haydenroche5 back to you for review, feedback addressed.  Once you're happy, please assign to @julek-wolfssl for a final review on the PKCS7 items if desired.

File: src/ssl.c (hunk 8 above)
Comment (haydenroche5) [reply src/ssl.c:62985]: Got it. I will add @julek-wolfssl to review this just for thoroughness.

**Review by haydenroche5 — APPROVED** [review]

File: src/ssl.c (hunk 8 above)
Comment (julek-wolfssl) [reply src/ssl.c:62985]: I didn't actually write these data structures. That's how it was when I was working on this section as well. I also was concerned about this solution but since `PKCS7 pkcs7` is the first field of the `WOLFSSL_PKCS7` struct, their addresses are guaranteed to match. This means that accessing the `PKCS7 pkcs7` field as either `(WOLFSSL_PKCS7*)p7->pkcs7.content` or as `(PKCS7*)p7->content` is safe. Obviously if the other fields were to be accessed when only `sizeof(PKCS7)` memory is allocated then it would result in a segfault. Overall I left this as is as I didn't see a real need to refactor this.

File: src/ssl.c [hunk 17]
```diff
@@ -62753,34 +62758,33 @@ WOLFSSL_STACK* wolfSSL_PKCS7_to_stack(PKCS7* pkcs7)
 WOLFSSL_STACK* wolfSSL_PKCS7_get0_signers(PKCS7* pkcs7, WOLFSSL_STACK* certs,
                                           int flags)
```
Comment (julek-wolfssl) [inline src/ssl.c:62759]: Ok so this function actually looks like a bug in OpenSSL itself. `get0` functions return objects that don't need to be `free`'d while the return of this function does need to be `free`'d. This is exactly what OpenSSL (up to latest master) does so its consistent but misleading.

File: src/ssl.c [hunk 18]
```diff
@@ -62915,13 +62919,333 @@ int wolfSSL_i2d_PKCS7_bio(WOLFSSL_BIO *bio, PKCS7 *p7)
...
+        if ((data = wolfSSL_BIO_new(wolfSSL_BIO_s_mem())) == NULL) {
+            WOLFSSL_MSG("Error in wolfSSL_BIO_new");
+            ret = 0;
+        }
+    }
+
+    /* prepend Content-Type header if PKCS7_TEXT */
+    if ((ret == 1) && (flags & PKCS7_TEXT)) {
+        if (wolfSSL_BIO_write(data, contTypeText,
+                              (int)XSTRLEN(contTypeText)) < 0) {
```
Comment (julek-wolfssl) [inline src/ssl.c:63172]: Use `XSTR_SIZEOF` to avoid extra calculations with `XSTRLEN`.

File: src/ssl.c [hunk 19]
```diff
@@ -62915,13 +62919,333 @@ int wolfSSL_i2d_PKCS7_bio(WOLFSSL_BIO *bio, PKCS7 *p7)
...
+        }
+        p7->data = (byte*)XMALLOC(memSz, NULL, DYNAMIC_TYPE_PKCS7);
+        if (p7->data == NULL) {
+            ret = 0;
+        }
+        else {
+            XMEMCPY(p7->data, mem, memSz);
+            p7->len = memSz;
+        }
+    }
```
Comment (julek-wolfssl) [inline src/ssl.c:63224]: Please use the following construct to extract the buffer (looking at our `BIO_free` we support `BIO_NOCLOSE` on a memory buffer so this should work):
```
 BUF_MEM *bptr;

 BIO_get_mem_ptr(mem, &bptr);
 BIO_set_close(mem, BIO_NOCLOSE); /* So BIO_free() leaves BUF_MEM alone */
 BIO_free(mem);
```
Taken from https://www.openssl.org/docs/manmaster/man3/BIO_get_mem_data.html

File: src/ssl.c [hunk 20]
```diff
@@ -62915,13 +62919,333 @@ int wolfSSL_i2d_PKCS7_bio(WOLFSSL_BIO *bio, PKCS7 *p7)
...
+
 int wolfSSL_PKCS7_verify(PKCS7* pkcs7, WOLFSSL_STACK* certs,
         WOLFSSL_X509_STORE* store, WOLFSSL_BIO* in, WOLFSSL_BIO* out, int flags)
 {
-    int ret = 0;
+    int i, ret = 0;
     unsigned char* mem = NULL;
     int memSz = 0;
     WOLFSSL_PKCS7* p7 = (WOLFSSL_PKCS7*)pkcs7;
+    static const char contTypeText[] = "Content-Type: text/plain\r\n\r\n";
```
Comment (julek-wolfssl) [inline src/ssl.c:63245]: This same buffer is used in two functions. Let's make this a single file-wide static buffer.

File: src/ssl.c [hunk 21]
```diff
@@ -63493,6 +63857,229 @@ WOLFSSL_API PKCS7* wolfSSL_SMIME_read_PKCS7(WOLFSSL_BIO* in,
...
+                WOLFSSL_MSG("Error in wc_InitRng");
+                ret = 0;
+            }
+
+            if ((ret > 0) && (wc_RNG_GenerateBlock(&rng, (byte*)boundary,
+                                  sizeof(boundary)) != 0)) {
+                    WOLFSSL_MSG("Error in wc_RNG_GenerateBlock");
+                    ret = 0;
+            }
+            wc_FreeRng(&rng);
```
Comment (julek-wolfssl) [inline src/ssl.c:63997]: Would it be safe to instead attempt to use the `globalRNG`? This doesn't look to be an operation that needs to be cryptographically secure.

File: src/ssl.c [hunk 22]
```diff
@@ -63493,6 +63857,229 @@ WOLFSSL_API PKCS7* wolfSSL_SMIME_read_PKCS7(WOLFSSL_BIO* in,
...
+        if (flags & PKCS7_DETACHED) {
+
+            /* generate random boundary */
+            if (wc_InitRng(&rng) != 0) {
+                WOLFSSL_MSG("Error in wc_InitRng");
+                ret = 0;
+            }
+
+            if ((ret > 0) && (wc_RNG_GenerateBlock(&rng, (byte*)boundary,
+                                  sizeof(boundary)) != 0)) {
```
Comment (julek-wolfssl) [inline src/ssl.c:63993]: `sizeof(boundary) - 1` (wasting one byte of entropy :sunglasses:)

File: src/ssl.c [hunk 23]
```diff
@@ -63493,6 +63857,229 @@ WOLFSSL_API PKCS7* wolfSSL_SMIME_read_PKCS7(WOLFSSL_BIO* in,
...
+                    WOLFSSL_MSG("Error in wc_RNG_GenerateBlock");
+                    ret = 0;
+            }
+            wc_FreeRng(&rng);
+
+            if (ret > 0) {
+                for (i = 0; i < (int)sizeof(boundary) - 1; i++) {
+                    boundary[i] =
+                        alphanum[boundary[i] % (sizeof(alphanum) - 1)];
```
Comment (julek-wolfssl) [inline src/ssl.c:64002]: Please use `XSTR_SIZEOF`.

File: src/ssl.c [hunk 24]
```diff
@@ -63493,6 +63857,229 @@ WOLFSSL_API PKCS7* wolfSSL_SMIME_read_PKCS7(WOLFSSL_BIO* in,
...
+    }
+    if (sigBase64 != NULL) {
+        XFREE(sigBase64, NULL, DYNAMIC_TYPE_TMP_BUFFER);
+    }
+
+    if (ret > 0) {
+        return 1;
+    }
+
+    return 0;
```
Comment (julek-wolfssl) [inline src/ssl.c:64080]: Please use `WOLFSSL_FAILURE` and `WOLFSSL_SUCCESS`.

**Review by julek-wolfssl — CHANGES_REQUESTED** [review]
Looks good overall. I'm curious why do you sometimes use `ret == 0` and sometimes `ret > 0` (`wolfSSL_SMIME_write_PKCS7`) for error fall through?

File: wolfcrypt/src/asn.c [hunk 25]
```diff
@@ -32301,19 +32301,22 @@ MimeParam* wc_MIME_find_param_attr(const char* attribute,
 * wc_MIME_canonicalize - Canonicalize a line by converting all line endings
 * to CRLF.
```
[the remark this answers is not in the document]
Comment (cconlon) [reply wolfcrypt/src/asn.c:32302]: Makes sense, will change to suggested naming.

File: wolfcrypt/src/asn.c [hunk 26]
```diff
@@ -32301,19 +32301,22 @@ MimeParam* wc_MIME_find_param_attr(const char* attribute,
...
     size_t end = 0;
     char* canonLine = NULL;
 
-    if (line == NULL || XSTRLEN(line) == 0) {
+    if (line == NULL || len == NULL || *len == 0) {
         return NULL;
     }
 
-    end = XSTRLEN(line);
+    end = *len - 1 ;
```
[the remark this answers is not in the document]
Comment (cconlon) [reply wolfcrypt/src/asn.c:32319]: Good catch, will adjust.

File: src/ssl.c [hunk 27]
```diff
@@ -62915,13 +62919,333 @@ int wolfSSL_i2d_PKCS7_bio(WOLFSSL_BIO *bio, PKCS7 *p7)
...
+            /* remove trailing null */
+            if (canonLine[canonLineLen] == '\0') {
+                canonLineLen--;
+            }
+
+            if (wolfSSL_BIO_write(out, canonLine, (int)canonLineLen) < 0) {
+                ret = -1;
+                break;
+            }
+            XFREE(canonLine, NULL, DYNAMIC_TYPE_TMP_BUFFER);
```
[the remark this answers is not in the document]
Comment (cconlon) [reply src/ssl.c:63104]: Good catch, will adjust.

File: src/ssl.c [hunk 28]
```diff
@@ -62915,13 +62919,333 @@ int wolfSSL_i2d_PKCS7_bio(WOLFSSL_BIO *bio, PKCS7 *p7)
...
+            WOLFSSL_MSG("Error prepending Content-Type header");
+            ret = 0;
+        }
+    }
+
+    /* convert line endings to CRLF if !PKCS7_BINARY */
+    if (ret == 1) {
+        if (flags & PKCS7_BINARY) {
+            /* no CRLF conversion, direct copy content */
+            if ((memSz = wolfSSL_BIO_get_mem_data(in, &mem)) < 0) {
```
[the remark this answers is not in the document]
Comment (cconlon) [reply src/ssl.c:63182]: Good improvement to not limit this to input memory BIOs, will switch to use `BIO_read()` instead.

File: tests/api.c [hunk 29]
```diff
@@ -46535,6 +46535,165 @@ static void test_wolfssl_PKCS7(void)
...
+        /* verify with wc_PKCS7_VerifySignedData */
+        AssertNotNull(p7Ver = wc_PKCS7_New(HEAP_HINT, devId));
+        AssertIntEQ(wc_PKCS7_Init(p7Ver, HEAP_HINT, INVALID_DEVID), 0);
+        AssertIntEQ(wc_PKCS7_VerifySignedData(p7Ver, out, outLen), 0);
+        wc_PKCS7_Free(p7Ver);
+
+        if (out != NULL) {
+            XFREE(out, NULL, DYNAMIC_TYPE_TMP_BUFFER);
+            out = NULL;
+        }
```
[the remark this answers is not in the document]
Comment (cconlon) [reply tests/api.c:46614]: Thanks, will adjust.

File: tests/api.c [hunk 30]
```diff
@@ -46535,6 +46535,165 @@ static void test_wolfssl_PKCS7(void)
...
+        PKCS7_free(p7);
+    }
+
+    /* TEST SUCCESS: Detached, not streaming, not MIME */
+    {
+        flags = PKCS7_BINARY | PKCS7_DETACHED;
+        AssertNotNull(p7 = PKCS7_sign(signCert, signKey, NULL, inBio, flags));
+        AssertIntGT((outLen = i2d_PKCS7(p7, &out)), 0);
+
+        /* verify with wolfCrypt, d2i_PKCS7 does not support detached content */
```
[the remark this answers is not in the document]
Comment (cconlon) [reply tests/api.c:46649]: Good idea, will add failure test.

File: tests/api.c [hunk 31]
```diff
@@ -46674,50 +46833,216 @@ static void test_wolfSSL_SMIME_read_PKCS7(void)
...
     AssertNotNull(pkcs7);
-    AssertIntEQ(wolfSSL_PKCS7_verify(pkcs7, NULL, NULL, bcont, NULL, PKCS7_NOVERIFY), SSL_SUCCESS);
+    out = wolfSSL_BIO_new(BIO_s_mem());
+    AssertNotNull(out);
+    AssertIntEQ(wolfSSL_PKCS7_verify(pkcs7, NULL, NULL, bcont, out,
+                PKCS7_NOVERIFY | PKCS7_TEXT), SSL_SUCCESS);
+    AssertIntGT((outBufLen = BIO_get_mem_data(out, &outBuf)), 0);
+    /* Content-Type should not show up in output buffer */
+    AssertIntGT(outBufLen, XSTRLEN(contTypeText));
+    AssertIntGT(XMEMCMP(outBuf, contTypeText, XSTRLEN(contTypeText)), 0);
```
[the remark this answers is not in the document]
Comment (cconlon) [reply tests/api.c:46900]: Sure, seems reasonable, I'll switch this.

File: src/ssl.c (hunk 17 above)
Comment (cconlon) [reply src/ssl.c:62759]: Good observation, I will add a comment note above this function mentioning that the returned stack needs to be freed.

File: src/ssl.c (hunk 18 above)
Comment (cconlon) [reply src/ssl.c:63172]: Thanks, will adjust.

File: src/ssl.c (hunk 19 above)
Comment (cconlon) [reply src/ssl.c:63224]: Sure, I can switch from `BIO_get_mem_data()` to `BIO_get_mem_ptr()`.  Doesn't seem to save much though - about the same line count, but one extra struct pointer added to the function.

File: src/ssl.c (hunk 20 above)
Comment (cconlon) [reply src/ssl.c:63245]: Sounds good, will adjust.

File: src/ssl.c (hunk 21 above)
Comment (cconlon) [reply src/ssl.c:63997]: Sure, seems like a good improvement.  Agreed, does not seem security critical.

File: src/ssl.c (hunk 22 above)
Comment (cconlon) [reply src/ssl.c:63993]: Right!  Will adjust.

File: src/ssl.c (hunk 23 above)
Comment (cconlon) [reply src/ssl.c:64002]: Will switch, thanks

File: src/ssl.c (hunk 24 above)
Comment (cconlon) [reply src/ssl.c:64080]: Will switch, thanks.

File: src/ssl.c (hunk 19 above)
Comment (julek-wolfssl) [reply src/ssl.c]: Using the way I posted above saves on the `malloc` and `memcpy`. You take control of the internal memory buffer so you can just assign it to `p7->data`. Actually this might not be possible with how we have the mallocs set up. Using BIO_get_mem_ptr would make the heap and type hints to not line up with the ones that will be used to free them. Do we have a way to know if the heap and type hints are actually used? If we do then we could avoid the `malloc` and `memcpy` but otherwise it will be best to leave it how you wrote it previously.

File: tests/api.c (hunk 31 above)
[the remark this answers is not in the document]
Comment (cconlon) [reply tests/api.c:46900]: I kept it as the XMEMCMP since outBuf isn't null terminated, and I think we really only care to check if it's at the beginning.

**Comment by cconlon** [comment]
Back to you @julek-wolfssl for re-review.

> I'm curious why do you sometimes use ret == 0 and sometimes ret > 0 (wolfSSL_SMIME_write_PKCS7)

In `wolfSSL_SMIME_write_PKCS7()` I chose to compare against `ret > 0` for error fall-through because the calls to `wolfSSL_BIO_write()` store a value > 0 into ret.  That allowed me not to convert the return from those function calls to a 0/1, and still be consistent with the `ret > 0` check throughout the function.

File: wolfcrypt/src/asn.c [hunk 32]
```diff
@@ -32557,22 +32557,25 @@ MimeParam* wc_MIME_find_param_attr(const char* attribute,
 }
 
 /*****************************************************************************
-* wc_MIME_canonicalize - Canonicalize a line by converting all line endings
-* to CRLF.
+* wc_MIME_canonicalize - Canonicalize a line by converting the trailing line
```
Comment (julek-wolfssl) [inline wolfcrypt/src/asn.c:32560]: `wc_MIME_canonicalize` -> `wc_MIME_single_canonicalize`

File: src/ssl.c [hunk 33]
```diff
@@ -62250,13 +62266,359 @@ int wolfSSL_i2d_PKCS7_bio(WOLFSSL_BIO *bio, PKCS7 *p7)
...
+            /* no line ending in current line, write direct to out */
+            if (wolfSSL_BIO_write(out, line, lineLen) < 0) {
+                ret = -1;
+                break;
+            }
+        }
+    }
+
+    if (canonLine != NULL) {
+        XFREE(canonLine, NULL, DYNAMIC_TYPE_TMP_BUFFER);
```
Comment (julek-wolfssl) [inline src/ssl.c:62464]: This should also use `DYNAMIC_TYPE_PKCS7`.

File: src/ssl.c [hunk 34]
```diff
@@ -62250,13 +62266,359 @@ int wolfSSL_i2d_PKCS7_bio(WOLFSSL_BIO *bio, PKCS7 *p7)
...
+            }
+            if (ret == 1) {
+                mem = (unsigned char*)XMALLOC(memSz, in->heap,
+                                              DYNAMIC_TYPE_TMP_BUFFER);
+                if (mem == NULL) {
+                    WOLFSSL_MSG("Failed to allocate memory for input data");
+                    ret = 0;
+                }
+                else {
+                    XMEMSET(mem, 0, memSz);
```
Comment (julek-wolfssl) [inline src/ssl.c:62544]: This `memset` isn't necessary. You write the entire buffer anyway.

File: tests/api.c (hunk 31 above)
[the remark this answers is not in the document]
Comment (julek-wolfssl) [reply tests/api.c:46900]: Lack of null termination is a deal breaker for `strstr`. `memcmp` sounds good then.

File: src/ssl.c (hunk 19 above)
Comment (julek-wolfssl) [reply src/ssl.c]: Actually please revert back to using `wolfSSL_BIO_get_mem_data`. Using `BIO_get_mem_ptr` won't save on a copy (as I originally thought it would) and with my memory BIO refactor in https://github.com/wolfSSL/wolfssl/pull/4956 it will be more correct to use `wolfSSL_BIO_get_mem_data`.

**Review by julek-wolfssl — CHANGES_REQUESTED** [review]
Just a few minor fixes and one remark in my last review should be reverted. Otherwise, looks great!

File: src/ssl.c (hunk 19 above)
Comment (cconlon) [reply src/ssl.c]: Sounds good, reverted back.

File: wolfcrypt/src/asn.c (hunk 32 above)
Comment (cconlon) [reply wolfcrypt/src/asn.c:32560]: Thanks, fixed

File: src/ssl.c (hunk 33 above)
Comment (cconlon) [reply src/ssl.c:62464]: Thanks, fixed

File: src/ssl.c (hunk 34 above)
Comment (cconlon) [reply src/ssl.c:62544]: Good catch, removed

**Review by julek-wolfssl — APPROVED** [review]
That looks to be everything. Nice work!
